spec/justified.tla

```diff
@@ -0,0 +1,209 @@
+    IN  LET 
+            \* @type: (Seq($targetMap), Int) => Seq($targetMap);
+            step(seq, i) == 
+                IF P(seq[1])
+                THEN seq
+                ELSE <<b(seq[1])>> \o seq \* Alternatively, we can append here and reverse the list at the end
+        IN ApaFoldSeqLeft( step, <<x>>, MkSeq(N, LAMBDA i: i) )
+        
+
+\* @type: ($targetMap, Int, $commonNodeState) => Set($checkpoint);
+AllJustifiedCheckpoints(initialTargetMap, N, node_state) ==
+    LET chain == Chain(initialTargetMap, N, node_state) IN
+    LET 
+        \* @type: ($targetMap, Set($checkpoint)) => Set($checkpoint);
+        G(currentTargetMap, justifiedCheckpoints) ==
+            LET targets == DOMAIN currentTargetMap IN
+            \* We define the justification filter for one checkpoint, which
+            \* looks at the currently computed set of justifiedCheckpoints
+            \* instead of recursing
+            LET 
+                \* @type: ($checkpoint) => Bool;
+                isOneCheckpointJustified(checkpoint) ==
+                    LET
+                        \* TODO: If we end up always doing 1-state model checking, these two are
+                        \* implied by the sate validity predicate and can be omitted
+                        hasBlockHash == has_block_hash(checkpoint.block_hash, node_state)
+                        isCompleteChain == 
+                            is_complete_chain(
+                                get_block_from_hash(checkpoint.block_hash, node_state), 
+                                node_state
+                            )
+                    IN  LET 
+                        \* TODO: Since GET_VALIDATOR_SET_FOR_SLOT is a constant operator, we could
+                        \* directly substitute it here
+                        validatorBalances == 
+                            GET_VALIDATOR_SET_FOR_SLOT(
+                                get_block_from_hash(checkpoint.block_hash, node_state), 
+                                checkpoint.block_slot, 
+                                node_state
+                            )
+                    IN LET 
+                        FFG_support_weight == 
+                            validator_set_weight(
+                                {
+                                    \* Instead of recursion, we take the 
+                                    \* votes relevant for this target, and filter them by known justified  sources
+                                    filtered_vote.sender : filtered_vote \in  {
+                                        vote \in currentTargetMap[checkpoint] : 
+                                            vote.message.ffg_source \in justifiedCheckpoints
+                                    }
+                                },
+                                validatorBalances
+                            )
+                        \* TODO: This value does not depend on any checkpoint or set of votes, so it is 
```

Comment:
I don't think this comment is correct. The value *does* depend on the checkpoint's slot (see `validatorBalances == GET_VALIDATOR_SET_FOR_SLOT(...)` above)

(though I don't know if the last parenthesis here refers to this first part, or the second one)

Reply:
Yes, I can see how this is poorly worded. What I meant to convey is that, at least for now, `GET_VALIDATOR_SET_FOR_SLOT ` is a constant operator, and if it stays as such, then `validatorBalances` is constant too.